cirq/circuits/circuit.py

```diff
@@ -68,74 +73,159 @@ class Circuit:
+                together. This option does not affect later insertions into the
+                circuit.
+            device: Hardware that the circuit should be able to run on.
+        """
+
+    @property
+    @abc.abstractmethod
+    def moments(self) -> Sequence['cirq.Moment']:
+        pass
+
+    @property
+    @abc.abstractmethod
+    def device(self) -> devices.Device:
+        pass
+
+    def freeze(self) -> 'cirq.FrozenCircuit':
+        '''Creates a FrozenCircuit from this circuit.
+
+        If 'self' is a FrozenCircuit, the original object is returned.
+        '''
+        from cirq.circuits import FrozenCircuit
+        if isinstance(self, FrozenCircuit):
+            return self
+        return FrozenCircuit(self,
+                             strategy=InsertStrategy.EARLIEST,
+                             device=self.device)
+
+    def unfreeze(self) -> 'cirq.Circuit':
+        '''Creates a Circuit from this circuit.
+
+        If 'self' is a Circuit, the original object is returned.
+        '''
+        if isinstance(self, Circuit):
+            return self
+        return Circuit(self,
+                       strategy=InsertStrategy.EARLIEST,
+                       device=self.device)
 
     def __bool__(self):
-        return bool(self._moments)
+        return bool(self.moments)
 
     def __eq__(self, other):
         if not isinstance(other, type(self)):
             return NotImplemented
-        return self._moments == other._moments and self._device == other._device
+        return self.moments == other.moments and self.device == other.device
 
     def _approx_eq_(self, other: Any, atol: Union[int, float]) -> bool:
         """See `cirq.protocols.SupportsApproximateEquality`."""
         if not isinstance(other, type(self)):
             return NotImplemented
         return cirq.protocols.approx_eq(
-            self._moments,
-            other._moments,
-            atol=atol
-        ) and self._device == other._device
+            self.moments, other.moments,
+            atol=atol) and self.device == other.device
 
     def __ne__(self, other) -> bool:
         return not self == other
 
     def __len__(self) -> int:
-        return len(self._moments)
+        return len(self.moments)
 
     def __iter__(self) -> Iterator['cirq.Moment']:
-        return iter(self._moments)
+        return iter(self.moments)
 
     def _decompose_(self) -> 'cirq.OP_TREE':
         """See `cirq.SupportsDecompose`."""
         return self.all_operations()
 
+    # pylint: disable=function-redefined
+    @overload
+    def __getitem__(self, key: int) -> 'cirq.Moment':
+        pass
+
+    @overload
+    def __getitem__(self, key: Tuple[int, 'cirq.Qid']) -> 'cirq.Operation':
+        pass
+
+    @overload
+    def __getitem__(self,
+                    key: Tuple[int, Iterable['cirq.Qid']]) -> 'cirq.Moment':
+        pass
+
+    @overload
+    def __getitem__(self, key: slice):
```

Comment:
Would be nice to add return type annotation here. We'd need a self-type annotation:
```python
Self = TypeVar('Self', bound='AbstractCircuit')
...

@overload
def __getitem__(self: Self, key: slice) -> Self:
    pass
```

Reply:
Is there a reason to prefer this structure over `def __getitem__(...) -> 'cirq.AbstractCircuit'`, which we've used for `Circuit` elsewhere in this file?

Reply:
I think it's nice to preserve that type information when slicing otherwise simple things may fail to type check like:
```python
c = cirq.Circuit(...)
c = c[:-1]  # error! cannot assign `AbstractCircuit` to variable of type `Circuit`
```
I totally made up that error message, but you get the idea :-)

Reply:
Ah, I see the difference now. Fix applied.